bchrpc/server.go

```diff
@@ -1113,16 +1208,421 @@ func (s *GrpcServer) GetMerkleProof(ctx context.Context, req *pb.GetMerkleProofR
+		meta := &pb.GetParsedSlpScriptResponse_V1Genesis{
+			V1Genesis: &pb.SlpV1GenesisMetadata{
+				Name:          slpMsg.Data.(v1parser.SlpGenesis).Name,
+				Ticker:        slpMsg.Data.(v1parser.SlpGenesis).Ticker,
+				DocumentUrl:   slpMsg.Data.(v1parser.SlpGenesis).DocumentURI,
+				DocumentHash:  slpMsg.Data.(v1parser.SlpGenesis).DocumentHash,
+				MintAmount:    slpMsg.Data.(v1parser.SlpGenesis).Qty,
+				MintBatonVout: uint32(slpMsg.Data.(v1parser.SlpGenesis).MintBatonVout),
+				Decimals:      uint32(slpMsg.Data.(v1parser.SlpGenesis).Decimals),
+			},
+		}
+		resp.SlpMetadata = meta
+		resp.Type = pb.SlpVersionType_SLP_V1_GENESIS
+	} else if slpMsg.TransactionType == "MINT" {
+		meta := &pb.GetParsedSlpScriptResponse_V1Mint{
+			V1Mint: &pb.SlpV1MintMetadata{
+				MintAmount:    slpMsg.Data.(v1parser.SlpMint).Qty,
+				MintBatonVout: uint32(slpMsg.Data.(v1parser.SlpMint).MintBatonVout),
+			},
+		}
+		resp.TokenId = slpMsg.Data.(v1parser.SlpMint).TokenID
+		resp.SlpMetadata = meta
+		resp.Type = pb.SlpVersionType_SLP_V1_MINT
+	} else if slpMsg.TransactionType == "SEND" {
+		meta := &pb.GetParsedSlpScriptResponse_V1Send{
+			V1Send: &pb.SlpV1SendMetadata{
+				Amounts: slpMsg.Data.(v1parser.SlpSend).Amounts,
+			},
+		}
+		resp.SlpMetadata = meta
+		resp.TokenId = slpMsg.Data.(v1parser.SlpSend).TokenID
+		resp.Type = pb.SlpVersionType_SLP_V1_SEND
+	}
+
+	return resp, nil
+}
+
+// GetTrustedSlpValidation returns SLP validity information about a specific token output
+func (s *GrpcServer) GetTrustedSlpValidation(ctx context.Context, req *pb.GetTrustedSlpValidationRequest) (*pb.GetTrustedSlpValidationResponse, error) {
+	if s.slpIndex == nil {
+		return nil, status.Error(codes.Unavailable, "slpindex required")
+	}
+
+	resp := &pb.GetTrustedSlpValidationResponse{}
+	results := make([]*pb.GetTrustedSlpValidationResponse_ValidityResult, len(req.Queries))
+	for i, query := range req.Queries {
+		result := &pb.GetTrustedSlpValidationResponse_ValidityResult{}
+		result.PrevOutHash = query.PrevOutHash
+		result.PrevOutVout = query.PrevOutVout
+
+		txid, err := chainhash.NewHash(query.PrevOutHash)
+		if err != nil {
+			return nil, status.Error(codes.Aborted, fmt.Sprintf("invalid txn hash (txo: %s:%s) - err %s", hex.EncodeToString(query.GetPrevOutHash()), string(query.GetPrevOutVout()), err.Error()))
+		}
+
+		entry, err := s.getSlpIndexEntry(txid)
+		if err != nil {
+			return nil, status.Error(codes.Aborted, "txid is missing from slp validity set (txo: "+hex.EncodeToString(query.GetPrevOutHash())+":"+string(query.GetPrevOutVout())+")")
+		}
+
+		if query.PrevOutVout == 0 || query.PrevOutVout > 19 {
+			return nil, status.Error(codes.Aborted, "slp output index cannot be 0 or > 19 (txo: "+hex.EncodeToString(query.GetPrevOutHash())+":"+string(query.GetPrevOutVout())+")")
+		}
+
+		slpMsg, err := v1parser.ParseSLP(entry.SlpOpReturn)
+
+		// set the proper slp version type
+		if slpMsg.TokenType == 0x01 {
```

Comment:
Yeah we definitely should have constants for each valid `TokenType`

Reply:
done